fdbserver/DataDistribution.actor.cpp

```diff
@@ -973,51 +1174,215 @@ struct DDTeamCollection {
+		//Remove the following commented code block later
+		/*
+		int i = 0;
+		std::map<Reference<TCMachineTeamInfo>, int, CompareTCMachineTeamInfoRef> usedMachineTeamMap;
+
+		for ( auto &machine : machine_info ) {
+			for ( auto &machineTeam : machine.second->machineTeams ) {
+				if ( usedMachineTeamMap.find(machineTeam) == usedMachineTeamMap.end() )
+					usedMachineTeamMap.insert(std::make_pair(machineTeam, 0));
+				else
+					usedMachineTeamMap[machineTeam]++;
+			}
+		}
+
+		assert( machineTeams.size() == usedMachineTeamMap.size() );
+
+		return usedMachineTeamMap.size();
+		*/
+	}
+
+	/**
+	 * Assume begin to end is sorted by std::sort
+	 * Assume InputIt is iterator to UID
+	 * @param begin: start of the team member ID
+	 * @param end: end of the team member ID
+	 * @param isIntialTeam: Is the team added at init() when we recreate teams by looking up DB
+	 */
+	template<class InputIt>
+	void addTeam( InputIt begin, InputIt end, bool isInitialTeam ) {
+		vector< Reference<TCServerInfo> > newTeamServers;
+
+		for ( auto i = begin; i != end; ++i ) {
+			if (server_info.find(*i) != server_info.end()) {
+				newTeamServers.push_back(server_info[*i]);
+			} else {
+				TraceEvent(SevWarn, "AddTeamServerNotFound").detail("ServerID", *i);
+			}
+		}
+
+		// We must allow creating empty teams because empty team is created when a remote DB is initialized.
+		// The empty team is used as the starting point to move data to the remote DB
+		if ( newTeamServers.empty() ) {
+			TraceEvent(SevWarn, "TeamCreation").detail("IsInitialTeam", isInitialTeam)
+				.detail("CreateEmptyTeam", 1).detail("Debug", "MayBeRight");
+		} else if ( newTeamServers.size() != configuration.storageTeamSize ) {
+			TraceEvent(SevWarn, "TeamCreation").detail("IsInitialTeam", isInitialTeam)
+				.detail("IncorrectTeamSize", newTeamServers.size()).detail("ExpectedTeamSize", configuration.storageTeamSize);
+		}
+
+		Reference<TCTeamInfo> teamInfo( new TCTeamInfo(newTeamServers) );
+		//TraceEvent(SevInfo, "TeamCreation", masterId).detail("IsInitialTeam", isInitialTeam)
+		//	.detail("Team", teamInfo->getDesc()).detail("ServerSize", newTeamServers.size());
+
+		if ( isInitialTeam == true ) {
+			//Initialize machine in case servers for initial team havenot set their machine
+			for ( auto i = begin; i != end; ++i ) {
+				if ( !server_info[*i]->machine.isValid() ) {
+					constructMachineFor1Server(*i);
+				}
+			}
+		}
+
+		//Add the reference of machineTeam (with machineIDs) into process team
+		vector< Standalone<StringRef> > machineIDs;
+		for ( auto i = begin; i != end; ++i ) {
+			machineIDs.push_back(server_info[*i]->machine->machineID);
+		}
+		sort(machineIDs.begin(), machineIDs.end());
+		Reference<TCMachineTeamInfo> machineTeamInfo = findMachineTeam(machineIDs);
+
+		if ( isInitialTeam && !machineTeamInfo.isValid() && !machineIDs.empty() ) { // Create machine team and assign to the server team
```

Comment:
I'm not sure if we should add the ASSERT(). 
After second thought, the ASSERT does not seem correct for me. 
The ASSERT should be
 ASSERT(isInitialTeam || machineTeamInfo.isValid());
 which uses isInitialTeam instead of !initialTeam